### PR TITLE
#425 use Solana's block height as neon block number

### LINKED ISSUE #425: Use solana's blockheight as block number instead of slot

### LINKED ISSUE #436: We need to rewrite Indexer partially as it works slower and slower depending on the number of transactions processed
The current version of Indexer works the following way:
- we need transaction number, which could be withdrawn by slot number
- Indexer loads all the slots from Solana
- parses them in the endless loop
- as the number of transactions grows, time of parsing grows as well
- in the result it's not possible to run a transaction at all

### PR DESCRIPTION
**Pull request by otselnik**

### CODE REVIEW

File: proxy/indexer/indexer_db.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        self.eth_sol_trx = SQLDict(tablename="ethereum_solana_transactions")
+        self.sol_eth_trx = SQLDict(tablename="solana_ethereum_transactions")
+        self.constants = SQLDict(tablename="constants")
+        if 'last_block_slot' not in self.constants:
+            self.constants['last_block_slot'] = 0
+
+
+    def submit_transaction(self, client, eth_trx, eth_signature, from_address, got_result, signatures):
+        (logs, status, gas_used, return_value, slot) = got_result
+        block_hash, block_number, _ = self.get_block_info(client, slot)
```
Comment (ivandzen) [inline proxy/indexer/indexer_db.py]: 'return None' case not handled

File: proxy/indexer/indexer.py [hunk 2]
```diff
@@ -444,83 +445,28 @@ def process_receipts(self):
...
+                height += 1
+            max_height = max(height_slot, key=height_slot.get)
+            max_slot = height_slot[max_height]
+            last_block = self.client._provider.make_request("getBlock", max_slot, {"commitment":"confirmed", "transactionDetails":"none", "rewards":False})['result']
+            if last_block['blockHeight'] == max_height:
+                self.db.set_last_block_slot(max_slot)
+                self.db.fill_block_height(height_slot)
+                last_block_slot = max_slot
+            else:
+                logger.debug(f"FAILED {max_height} {max_slot} {last_block}")
```
Comment (ivandzen) [inline proxy/indexer/indexer.py]: logger.warning/error is better to be used here

File: proxy/indexer/indexer_db.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+            for idx, sig in enumerate(signatures):
+                self.sol_eth_trx[sig] = {
+                    'idx': idx,
+                    'eth': eth_signature,
+                }
+            return True
+        return False
+
+    def get_block_info(self, client, slot):
+        block = self.blocks.get(slot, None)
```
Comment (ivandzen) [inline proxy/indexer/indexer_db.py]: I don't see where data stored to this container. Seems it will always be empty

**Review by ivandzen — CHANGES_REQUESTED** [review]

File: proxy/indexer/indexer_db.py (hunk 3 above)
Comment (ivandzen) [reply proxy/indexer/indexer_db.py]: Added fix (see below line 87)

File: proxy/indexer/indexer_db.py (hunk 1 above)
Comment (ivandzen) [reply proxy/indexer/indexer_db.py]: Fix added

File: proxy/indexer/indexer.py (hunk 2 above)
Comment (ivandzen) [reply proxy/indexer/indexer.py]: Fix added

**Comment by ivandzen** [comment]
Suggest to add some tests

File: proxy/plugin/solana_rest_api.py [hunk 4]
```diff
@@ -229,29 +222,33 @@ def eth_getStorageAt(self, account, position, block_identifier):
...
         return ret
 
     def eth_getBlockByNumber(self, tag, full):
         """Returns information about a block by block number.
             tag - integer of a block number, or the string "earliest", "latest" or "pending", as in the default block parameter.
             full - If true it returns the full transaction objects, if false only the hashes of the transactions.
         """
-        slot = self.process_block_tag(tag)
+        block_number = self.process_block_tag(tag)
+        slot = self.db.get_slot_by_number(block_number)
```
Comment (afalaleev) [inline proxy/plugin/solana_rest_api.py]: _eth_getBlockByNumber('latest')_ on a live Solana cluster (testnet/mainnet/devnet) will return many errors.
Reasons:
- _process_block_tag('latest')_ returns the last block on the Solana node
- _db.get_slot_by_number(height)_ returns the last block in the DB

So, if block doesn't exist in DB,  a clients receive error _"Not found block by number latest"_

**Review by afalaleev — CHANGES_REQUESTED** [review]

File: proxy/indexer/indexer.py [hunk 5]
```diff
@@ -444,83 +445,28 @@ def process_receipts(self):
...
-                block_hash = '0x' + base58.b58decode(block['blockhash']).hex()
-                retry = False
-            except Exception as err:
-                logger.debug(err)
-                time.sleep(1)
-
-        return (slot, block_hash)
+        start_time = time.time()
+        last_block_slot = self.db.get_last_block_slot()
+        confirmed_blocks = self.client.get_confirmed_blocks(last_block_slot)["result"]
```
Comment (afalaleev) [inline proxy/indexer/indexer.py]: _get_confirmed_blocks()_ is deprecated method.
And it returns only 500'000 blocks.

**Review by afalaleev — CHANGES_REQUESTED** [review]

**Review by afalaleev — APPROVED** [review]

**Pull request merged**